File: app/src/main/java/dev/n3shemmy3/coffre/backend/dao/TransactionDao.java
package dev.n3shemmy3.coffre.backend.dao;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Update;

import java.math.BigDecimal;
import java.util.List;

import dev.n3shemmy3.coffre.backend.item.Transaction;

@Dao
public interface TransactionDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insert(Transaction item);

    @Update
    void update(Transaction item);

    @Query("DELETE FROM " + Transaction.TABLE_NAME + " WHERE id = :id")
    int delete(long id);

    @Query("SELECT SUM(CASE WHEN transactionType = 'INCOME' THEN amount ELSE -amount END) FROM " + Transaction.TABLE_NAME)
    LiveData<BigDecimal> getNetBalance();

    @Query("SELECT SUM(amount) FROM " + Transaction.TABLE_NAME + " WHERE transactionType = 'INCOME'")
    LiveData<BigDecimal> getTotalIncome();

    @Query("SELECT SUM(amount) FROM " + Transaction.TABLE_NAME + " WHERE transactionType = 'EXPENSE'")
    LiveData<BigDecimal> getTotalExpenses();

    @Query("DELETE FROM " + Transaction.TABLE_NAME)
    void deleteAllTransactions();

    @Query("SELECT * FROM " + Transaction.TABLE_NAME)
    LiveData<List<Transaction>> getAllTransactions();

    @Query("SELECT * FROM " + Transaction.TABLE_NAME + " WHERE title LIKE '%' || :query || '%'")
    LiveData<List<Transaction>> search(String query);

}

File: app/src/main/kotlin/dev/n3shemmy3/coffre/data/dao/TransactionDao.kt
/*
 *
 *  * Copyright (C) 2025 Shemmy
 *  *
 *  * This program is free software: you can redistribute it and/or modify
 *  * it under the terms of the GNU General Public License as published by
 *  * the Free Software Foundation, version 3 of the License.
 *  *
 *  * This program is distributed in the hope that it will be useful,
 *  * but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  * GNU General Public License for more details.
 *  *
 *  * You should have received a copy of the GNU General Public License
 *  * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 */

package dev.n3shemmy3.coffre.data.dao

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import androidx.room.Update
import dev.n3shemmy3.coffre.data.entity.Transaction
import kotlinx.coroutines.flow.Flow
import java.math.BigDecimal

@Dao
interface TransactionDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insert(transaction: Transaction)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insert(transactions: List<Transaction>)

    @Update
    suspend fun update(transaction: Transaction)

    @Update
    suspend fun update(transactions: List<Transaction>)

    @Query("DELETE FROM transactions WHERE id = :id")
    suspend fun deleteById(id: Long)

    @Delete
    suspend fun delete(transactions: List<Transaction>)

    @Query("SELECT * FROM transactions WHERE id = :id LIMIT 1")
    suspend fun getById(id: Long): Transaction?

    @Query("SELECT * FROM transactions WHERE id IN (:ids)")
     fun getByIds(ids: List<Long>): Flow<List<Transaction>>

    @Query("SELECT * FROM transactions")
    fun getAll(): Flow<List<Transaction>>

    @Query("SELECT COALESCE(SUM(amount), 0) FROM transactions WHERE type = 'INCOME'")
    fun getTotalIncomes(): Flow<BigDecimal>

    @Query("SELECT COALESCE(SUM(amount), 0) FROM transactions WHERE type = 'EXPENSE'")
    fun getTotalExpenses(): Flow<BigDecimal>

    @Query("SELECT COALESCE(SUM(CASE WHEN type='INCOME' THEN amount ELSE -amount END), 0) FROM transactions")
    fun getTotalBalance(): Flow<BigDecimal>

}
